spring-cloud-kubernetes-core/src/main/java/io/fabric8/spring/cloud/kubernetes/reload/ConfigurationChangeDetector.java

```diff
@@ -0,0 +1,109 @@
+package io.fabric8.spring.cloud.kubernetes.reload;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.PreDestroy;
+
+import io.fabric8.kubernetes.client.Client;
+import io.fabric8.kubernetes.client.KubernetesClient;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.env.CompositePropertySource;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.MapPropertySource;
+import org.springframework.core.env.PropertySource;
+
+/**
+ * This is the superclass of all beans that can listen to changes in the configuration and fire a reload.
+ */
+public abstract class ConfigurationChangeDetector {
+
+    protected Logger log = LoggerFactory.getLogger(getClass());
+
+    protected ConfigurableEnvironment environment;
+
+    protected ConfigReloadProperties properties;
+
+    protected KubernetesClient kubernetesClient;
+
+    protected ConfigurationUpdateStrategy strategy;
+
+    public ConfigurationChangeDetector(ConfigurableEnvironment environment, ConfigReloadProperties properties, KubernetesClient kubernetesClient, ConfigurationUpdateStrategy strategy) {
+        this.environment = environment;
+        this.properties = properties;
+        this.kubernetesClient = kubernetesClient;
+        this.strategy = strategy;
+    }
+
+    @PreDestroy
+    public void shutdown() {
+        // Ensure the kubernetes client is cleaned up from spare threads when shutting down
+        if (kubernetesClient instanceof Client) {
```

Comment:
KubernetesClient extends Client, so this will always evaluate true.
So, you can just call close() on the `kubernetesClient` object without casting etc.